Src/zipkin4net/Src/ISpan.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+
+namespace zipkin4net
+{
+    public enum SpanKind
+    {
+        Client,
+        Server,
+
+        /// <summary>
+        /// When present, <see cref="Start"/> is the moment a consumer sent a message to a destination. A
+        /// duration between <see cref="Start"/> and <see cref="Finish"/> may imply batching delay.
+        /// <see cref="RemoteEndpoint(endpoint)"/> indicates the origin, such as a broker.
+        ///
+        /// <p>Unlike <see cref="Client"/>, messaging spans never share a span ID. For example, the
+        /// <see cref="Consumer"/> of the same message has <see cref="ITraceContext.ParentSpanId"/> set to this span's
+        /// <see cref="ITraceContext.SpanId"/>.</p>
+        /// </summary>
+        Producer,
+
+        /// <summary>
+        /// When present, <see cref="Start"/> is the moment a consumer received a message from an origin. A
+        /// duration between <see cref="Start"/> and <see cref="Finish"/> may imply a processing backlog. while
+        /// <see cref="RemoteEndpoint(endpoint)"/> indicates the origin, such as a broker.
+        ///
+        /// <p>Unlike <see cref="Server"/>, messaging spans never share a span ID. For example, the
+        /// <see cref="Producer"/> of this message is the <see cref="ITraceContext.ParentSpanId"/> of this span.</p>
+        /// </summary>
+        Consumer
+    }
+
+    ///  <summary>
+    ///  Used to model the latency of an operation.
+    ///  <p>For example, to trace a local function call.</p>
+    ///  <pre><code>
+    ///  Span span = tracer.NewTrace().Name("encode").Start();
+    ///  try {
+    ///    doSomethingExpensive();
+    ///  } finally {
+    ///    span.Finish();
+    ///  }
+    ///  </code>
+    ///  </pre>
+    ///  This captures duration of <see cref="M:zipkin4net.ISpan.Start(System.Int64)" /> until <see cref="M:zipkin4net.ISpan.Finish(System.Int64)" /> is called.
+    ///  </summary>
+    //TODO internal for now. It's still work in progress
+    internal interface ISpan : ISpanCustomizer
+    {
+        ITraceContext Context { get; }
+
+        /// <summary>
+        /// The kind of span is optional. When set, it affects how a span is reported. For example, if the
+        /// kind is <see cref="SpanKind.Server"/>, the span's start timestamp is implicitly annotated as "sr" and
+        /// that plus its duration as "ss".
+        /// </summary>
+        ISpan Kind(SpanKind kind);
+
+        /// <summary>
+        /// Starts the span with an implicit timestamp.
+        /// 
+        /// <p>Spans can be modified before calling start. For example, you can add tags to the span and
+        /// set its name without lock contention.</p>
+        /// </summary>
+        /// <returns></returns>
+        ISpan Start();
+
+        /// <summary>
+        /// Like <see cref="Start()"/>, except with a given timestamp in microseconds.
+        /// </summary>
+        /// <returns></returns>
+        ISpan Start(DateTime timestamp);
```

Comment:
we've had some chat about this method... as long as a reporter is available maybe best to leave out the start with pre-ordained timestamp function, as most of the time pre-ordained timestamps mean you are recording after the fact (and that implies you can just use a span reporter). If not the case, here, for example other code still needs this api, sounds fine just yeah..

Reply:
actually this ends up used in subtle cases